packages/react-jss/src/createUseStyles.js

```diff
@@ -0,0 +1,106 @@
+// @flow
+
+import React from 'react'
+import {ThemeContext as DefaultThemeContext} from 'theming'
+
+import JssContext from './JssContext'
+import {
+  createStaticSheet,
+  addDynamicRules,
+  removeDynamicRules,
+  updateDynamicRules
+} from './utils/sheets'
+import {getIndex} from './utils/index-counter'
+import type {HookOptions, Styles} from './types'
+import {unmanageSheet, manageSheet} from './utils/managers'
+import {getSheetClasses} from './utils/get-sheet-classes'
+
+const noTheme = {}
+
+const createUseStyles = <Theme: {}>(styles: Styles<Theme>, options?: HookOptions<Theme> = {}) => {
+  const {index = getIndex(), theming, name = 'Hook', ...sheetOptions} = options
+  const ThemeContext = (theming && theming.context) || DefaultThemeContext
+  const useTheme =
+    typeof styles === 'function'
+      ? // $FlowFixMe
+        (): Theme => React.useContext(ThemeContext)
+      : // $FlowFixMe
+        (): Theme => noTheme
+
+  return (data: any) => {
+    const context = React.useContext(JssContext)
+    const theme = useTheme()
+
+    // When the theme or the context changes we create a new sheet
+    const sheet = React.useMemo(
+      () => {
+        if (context.disableStylesGeneration) {
+          return undefined
+        }
+
+        const staticSheet = createStaticSheet({
```

Comment:
To reiterate on the render vs. commit phase leak: This can potentially be called multiple times after/before effect cleanups are executed.

It should ensure that unmanage and manage happens during render and only DOM manipulations happen in useLayoutEffect. It seems like sheet managing is coupled to the DOM currently in Jss?

Reply:
Sheet manager https://github.com/cssinjs/jss/blob/master/packages/jss/src/SheetsManager.js
has 2 functions
1. count the refs - the amount of component instances using a sheet
2. calling attach/detach once first component is using a sheet or none is using

Since .attach/.detach bascially calls the DomRenderer functions which do nothing if we are not on the client, we don't have a strong coupling to the DOM.

Reply:
> Since .attach/.detach bascially calls the DomRenderer functions

If there is no manage/unmanage functionality that does not also use the DOM then it is tightly coupled. Just because something is a no-op in some environment doesn't mean it doesn't create a coupling. 

If you want to be concurrent ready then this functionality needs to be split up. There will be a lot of issues in concurrent react if you have side-effects in render that are not cleaned up in the same phase.

Reply:
cleaning up can be done if we need it, but I don't know for sure the exact requirements there, so why bother now? It seems like something we can solve later, no?

Reply:
Maybe. I think pre-release would be a good chance to get this right since we already know that |render| >= |commit| and we have ways to test that our solutions work with that assumptions (stable: StrictMode, unstable: unstable_ConcurrentMode).

I'm currently creating some sandboxes that compare different styling solutions in different modes. I would hope that a styling solution using hooks does not introduce regressions.

Reply:
As we have a lot of confusion around what needs to happen, here is a breakdown:

On SSR:
Create the sheet and add it to the registry. We would also need to add the dynamic rules and update them.

On the client:
Create sheet, add the dynamic rules and update them. Here we also need to call so `manage` the sheets which will attach them to the DOM.

We are currently adding the sheet to the registry in the managing phase. So by splitting them up, we could solve the SSR problem very quickly.

My question would be: What happens if we call `setState` inside a layout effect? Will the state update be reflected in the first render?
Because on the client we also need to think about theme updates aka recreating the sheet. I would assume this should go into the a layout effect, but this would mean we would also need to move the initial creation to a layout effect

Reply:
> Create the sheet and add it to the registry. We would also need to add the dynamic rules and update them.

Are you sure we need this on SSR? It is supposed to be a single pass, rules should only need the initial state, without component updates.

Reply:
> Create sheet, add the dynamic rules and update them. Here we also need to call so manage the sheets which will attach them to the DOM.

Currently SheetsRegistry.toString() only returns attached nodes AFAIK. That's why we also call a .attach() during SSR. This basically marks sheets that we need to output during SSR, because there will be more created than mounted. Seems like we need a change in this logic.

Reply:
> My question would be: What happens if we call setState inside a layout effect? Will the state update be reflected in the first render?

Please verify this, but afaik  - no, SSR is a single pass.